cmd/mdatagen/validate.go

```diff
@@ -129,34 +129,9 @@ func (md *metadata) validateResourceAttributes() error {
 func (md *metadata) validateMetrics() error {
 	var errs error
 	usedAttrs := map[attributeName]bool{}
-	for mn, m := range md.Metrics {
-		if m.Sum == nil && m.Gauge == nil {
-			errs = errors.Join(errs, fmt.Errorf("metric %v doesn't have a metric type key, "+
-				"one of the following has to be specified: sum, gauge", mn))
-			continue
-		}
-		if m.Sum != nil && m.Gauge != nil {
-			errs = errors.Join(errs, fmt.Errorf("metric %v has more than one metric type keys, "+
-				"only one of the following has to be specified: sum, gauge", mn))
-			continue
-		}
-		if err := m.validate(); err != nil {
-			errs = errors.Join(errs, fmt.Errorf(`metric "%v": %w`, mn, err))
-			continue
-		}
-		unknownAttrs := make([]attributeName, 0, len(m.Attributes))
-		for _, attr := range m.Attributes {
-			if _, ok := md.Attributes[attr]; ok {
-				usedAttrs[attr] = true
-			} else {
-				unknownAttrs = append(unknownAttrs, attr)
-			}
-		}
-		if len(unknownAttrs) > 0 {
-			errs = errors.Join(errs, fmt.Errorf(`metric "%v" refers to undefined attributes: %v`, mn, unknownAttrs))
-		}
-	}
-	errs = errors.Join(errs, md.validateAttributes(usedAttrs))
+	errs = errors.Join(errs, validateMetrics(md.Metrics, md.Attributes, usedAttrs),
+		validateMetrics(md.Telemetry.Metrics, md.Attributes, usedAttrs),
```

Comment:
Thanks @Frapschen, is there already a unit test that covers this change?

Reply:
Yes, please see the changed file `cmd/mdatagen/testdata/unused_attribute.yaml`, it will make sure the attributes that are used in `telemetry.metric` will not encounter `unused attributes: [xxx]` error.